go/vt/sidecardb/sidecardb.go

```diff
@@ -0,0 +1,406 @@
+	_, err := si.exec(si.ctx, CreateVTDatabaseQuery, 1, false, false)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+	log.Infof("createVTDatabase: %s", CreateVTDatabaseQuery)
+	return nil
+}
+
+// sets db of current connection, returning the currently selected database
+func (si *vtSchemaInit) setCurrentDatabase(dbName string) (string, error) {
+	rs, err := si.exec(si.ctx, SelectCurrentDatabaseQuery, 1, false, false)
+	if err != nil {
+		return "", err
+	}
+	if rs == nil || rs.Rows == nil { // we get this in tests
+		return "", nil
+	}
+	currentDB := rs.Rows[0][0].ToString()
+	if currentDB != "" { // while running tests we can get currentDB as empty
+		_, err = si.exec(si.ctx, fmt.Sprintf("use %s", dbName), 1000, false, false)
+		if err != nil {
+			return "", err
+		}
+	}
+	return currentDB, nil
+}
+
+// gets existing schema of table
+func (si *vtSchemaInit) getCurrentSchema(tableName string) (string, error) {
+	var currentTableSchema string
+
+	rs, err := si.exec(si.ctx, fmt.Sprintf(ShowCreateTableQuery, tableName), 1, false, false)
+	if err != nil {
+		log.Errorf("Error getting _vt table schema for %s: %+v", tableName, err)
+		return "", err
+	}
+	if len(rs.Rows) > 0 {
+		currentTableSchema = rs.Rows[0][1].ToString()
+	}
+	return currentTableSchema, nil
+}
+
+// remove the extra DEFAULT CHARSET which we get from "show create table"
+func stripCharset(schema string) string {
+	ind := strings.Index(schema, "DEFAULT CHARSET")
+	if ind <= 0 {
+		return schema
+	}
+	return schema[:ind]
+}
+
+// finds diff that needs to be applied to current table schema to get the desired one. Will be an empty string if they match.
+// could be a create statement if the table does not exist or an alter if table exists but has a different schema
+func (si *vtSchemaInit) findTableSchemaDiff(current, desired string) (string, error) {
+	// todo: remove once this is fixed in schemadiff
+	current = stripCharset(current)
+
+	hints := &schemadiff.DiffHints{}
+	diff, err := schemadiff.DiffCreateTablesQueries(current, desired, hints)
+	if err != nil {
+		return "", err
+	}
+
+	var tableAlterSQL string
+	if diff != nil {
+		tableAlterSQL = diff.CanonicalStatementString()
+
+		// temp logging, should be removed in v17
+		log.Infof("alter sql %s", tableAlterSQL)
+		log.Infof("current schema %s", current)
+	}
+
+	return tableAlterSQL, nil
+}
+
+// createOrUpgradeTable expects that we are already in the _vt database
+func (si *vtSchemaInit) createOrUpgradeTable(table *vtTable) error {
+	var desiredTableSchema string
+	ctx := si.ctx
+	bytes, err := schemaLocation.ReadFile(table.path)
+	if err != nil {
+		return err
+	}
+	desiredTableSchema = string(bytes)
+
+	var tableAlterSQL string
+	tableExists := si.tableExists(table.name)
+	if tableExists {
+		currentTableSchema, err := si.getCurrentSchema(table.name)
+		if err != nil {
+			return err
+		}
+		tableAlterSQL, err = si.findTableSchemaDiff(currentTableSchema, desiredTableSchema)
+		if err != nil {
+			return err
+		}
```

Comment:
schemadiff doesn't read from the database: it needs to be given both versions of the table: the existing one and the desired one. 

What is likely to slow down the init on existing tablets?  We don't run any DDLs if tables exist and schemadiff finds no differences. For initialized tablets, on a restart, we will run 1 select query to get list of all _vt tables, one `show create table` for each table (currently 18) and  schemadiff won't signal changes. In my e2e test run locally, it takes 54ms to create all tables on new and 7ms on an already inited tablet.

Reply:
I agree that `if not exists` should not be needed. I started without it and then added it when I got some errors in some paths like backup restore on replicas where the database already had the tables and we also found these `create`s in the binlog and the replica tablet would halt since it could not execute the duplicate create. 

Now I know this **should never happen**, but as an abundance of caution I have left it so that we don't have clusters that don't come up. We can remove it in v17 or v18 once the schema init is stable. I have added notes about this in `sidecardb/doc.go`.